type-auto-algorithm.md

```diff
@@ -0,0 +1,57 @@
+# `--type=auto` Module Type Detection Algorithm
+
+This is the algorithm, in psuedocode, for the `--type=auto`/`-a` Node startup option described [here](./README.md). It is based on detailed syntax rules outlined [here](https://github.com/nodejs/node-eps/issues/57#issuecomment-300870976).
+
+This algorithm relies on the fact that it is evaluating an initial entry point—the very first file loaded for a program. Affirmatively detecting CommonJS would be virtually impossible otherwise, as it would no longer be a safe assumption that a reference to a globally defined `require` confirms that a file is CommonJS (since a file loaded earlier in the program could have defined `require` in the global scope). Additionally, many files (either CommonJS or ESM) that are not entry points are often ambiguous, for example containing only `console.log` statements or references to user-defined globals. For this algorithm to be both confident (no false positives) and inclusive (few files judged ambiguous) it must be restricted to initial entry points.
+
+## Step 1: Parse
+
+The initial entry point file must be parsed so that its syntax may be analyzed.
+
+## Step 2: Look for signifiers of ESM JavaScript
+
+If the source code contains either of the following:
+
+- An `import` statement, like `import "x";`
+
+- An `export` statement, like `export {};`
+
+Then the file is unambiguously ESM. Stop and execute it as ESM JavaScript.
+
+These are the only two syntactical elements that are unambiguously ESM. The `import` expression, `import("x")`, is allowed in CommonJS.
+
+If the source code did not contain an ESM signifier, continue:
+
+## Step 3: Look for signifiers of CommonJS JavaScript
+
+If the source code contains any of the following:
+
+- A reference to a variable named `__filename` without a declaration for that variable.
+
+- A reference to a variable named `__dirname` without a declaration for that variable.
+
+- A reference to a variable named `require` without a declaration for that variable.
+
+- A reference to a variable named `module` without a declaration for that variable.
+
+- A reference to a variable named `exports` without a declaration for that variable.
+
+- A reference to `arguments` in the top scope.
+
+- A reference to `return` in the top scope.
```

Comment:
this ain't a reference

Reply:
What is it?

Reply:
Maybe:

> - A `return` statement in the top scope.

Reply:
If we want to be very specific:
* A `return` statement in the top VariableEnvironment of the source text

Reply:
I'm assuming this applied to returns within things like block statements.

Reply:
So really it's a `return` statement in the top function scope?

Reply:
Not a function scope necessarily. So no, not a function scope.

Reply:
It's not? Doesn't the CommonJS wrapper create a function scope that encompasses the entire file? So a `return` anywhere in there would exit that wrapper function scope, unless the `return` were inside a child function scope?

Reply:
But that is circular, you must treat it as CJS for you to check if it is CJS in that case. Nothing is preventing Module (or god forbid Script) from adding early `return`s except going through committee. Same for `await`. You should state the heuristics based upon the overlap, which would just be the VariableEnvironment, which is the Module environment for Modules, the Global environment for Script, and the Function environment for CJS. Only stating it on the function environment doesn't make sense since this is just a heuristic, we cannot assert that it is a function environment in order to check if it is a function environment.

Reply:
Is it really that common to use `return` in the top scope (or whatever I should call it) _without_ the CommonJS globals? I'm wondering if we need to bother with the checks beyond looking for use of those globals (including checking that the globals aren't defined by user code).

If we limit the CommonJS check to just the globals, then the explanation for this feature is straightforward: import/export is run as ESM, use of CommonJS globals is run as CommonJS, anything else errors. The simplicity of that might be worth a few false negatives (my "anything else" group) if these other checks are unlikely to grow our haul of detected CommonJS significantly.